### PR TITLE
Add feature: Whitelist. 

### PR DESCRIPTION
**Pull request by tramhao**
Optional only scan songs under start folder and default to false. 
Can be used combined with blacklist.

### CODE REVIEW

**Comment by tramhao** [comment]
I think this feature is quite useful as I always have many media files from other packages in my phone and generally speaking I don't need to play them. I saw a related issue was closed before due to no progress.https://github.com/AdrienPoupa/VinylMusicPlayer/issues/101

File: app/src/main/java/com/poupa/vinylmusicplayer/discog/Discography.java
Comment (soncaokim) [inline app/src/main/java/com/poupa/vinylmusicplayer/discog/Discography.java:294]: Please either rename the `isBlacklisted` predicate, or create a new one (ie. `isIncluded = isWhitelisted && !isBlacklisted`) to express clearly the intention.

File: app/src/main/res/xml/pref_blacklist.xml
Comment (soncaokim) [inline app/src/main/res/xml/pref_blacklist.xml:12]: My understanding that white list takes precedence over black list (the track must be in white list first, then will be excluded if in black list).

If this is correct:
- I'd put the corresponding pref before the black list pref.
- and rename the "white list" to "Limit to start folder" (ie white list doesnt seem to be the right wording).
- optionally, rename "black list" to "Exclude content from folder"

File: app/src/main/res/values/strings.xml
Comment (soncaokim) [inline app/src/main/res/values/strings.xml:364]: "only scan..." -> "only show..."

**Review by soncaokim — COMMENTED** [review]
Thanks for the submission - the idea looks okay, with few comments I've left inline.
I'll play with the UI during the weekend and provide additional feedback.

**Comment by soncaokim** [comment]
If white list pref is changed, you'll need to trigger a rescan - with confirmation popup or a toast message explaining that it might take time to complete.

**Comment by soncaokim** [comment]
Extra requests, on the visual presentation (@AdrienPoupa please  share your opinion on this): 

- I'd prefer merging the Setting's blacklist section (the existing blacklist and the new whitelist) into the Library section (on top).
  These are options to control the content/presentation of the library.
- Show the current value of start folder as the description/subtitle of the new whitelist setting (ie "only show content under the start folder: /storage/music/xxx")

File: app/src/main/res/xml/pref_blacklist.xml
Comment (tramhao) [reply app/src/main/res/xml/pref_blacklist.xml:12]: > My understanding that white list takes precedence over black list (the track must be in white list first, then will be excluded if in black list).
> 
> If this is correct:
> 
> * I'd put the corresponding pref before the black list pref.
> * and rename the "white list" to "Limit to start folder" (ie white list doesnt seem to be the right wording).
> * optionally, rename "black list" to "Exclude content from folder"

Yes, this is the idea. The actual meaning is "Limit to start folder". But blacklist and whitelist name seems more straight-forward. If you think it's better to change, please comment and I'll change.

**Comment by tramhao** [comment]
> If white list pref is changed, you'll need to trigger a rescan - with confirmation popup or a toast message explaining that it might take time to complete.

Yes, would you please point me to the correct way to do it? I'm new to Android and Vinyl is a great app. Thanks.

**Comment by soncaokim** [comment]
> > If white list pref is changed, you'll need to trigger a rescan - with confirmation popup or a toast message explaining that it might take time to complete.
> Yes, would you please point me to the correct way to do it?

This would be:
- [ ] listen to the change of pref, i.e implement `SharedPreferences.OnSharedPreferenceChangeListener` on the class that manages the library. Either in Discography or create a new WhitelistManager class.
- [ ] translate it as a notification to the rest of the app, like this one: https://github.com/AdrienPoupa/VinylMusicPlayer/blob/28dd86923c6faa10cb800ffec1224569a102dd9e/app/src/main/java/com/poupa/vinylmusicplayer/provider/BlacklistStore.java#L129
- [x] already in place: this notification will be handled by different components in the app, including Discog/views/etc

**Comment by soncaokim** [comment]
> Either in Discography or create a new WhitelistManager class.

I'd prefer to have the pref change listener outside of Discog class, to reduce the dependency of this class on the user UI part.

**Comment by tramhao** [comment]
> > > If white list pref is changed, you'll need to trigger a rescan - with confirmation popup or a toast message explaining that it might take time to complete.
> > 
> > 
> > Yes, would you please point me to the correct way to do it?
> 
> This would be:
> 
> * [ ]  listen to the change of pref, i.e implement `SharedPreferences.OnSharedPreferenceChangeListener` on the class that manages the library. Either in Discography or create a new WhitelistManager class.
> * [ ]  translate it as a notification to the rest of the app, like this one: https://github.com/AdrienPoupa/VinylMusicPlayer/blob/28dd86923c6faa10cb800ffec1224569a102dd9e/app/src/main/java/com/poupa/vinylmusicplayer/provider/BlacklistStore.java#L129
> * [x]  already in place: this notification will be handled by different components in the app, including Discog/views/etc

Thanks for the help. Follow your idea, I found that musicservice.java implemented OnSharedPreferenceChangeListener. Now it's working, but is it the appropriated place to do it? Thanks.

**Comment by tramhao** [comment]
SettingsActivity.java also implements OnSharedPreferenceChangeListener and seems a better place to intrigue the rescan.

File: app/src/main/java/com/poupa/vinylmusicplayer/service/MusicService.java
Comment (soncaokim) [inline app/src/main/java/com/poupa/vinylmusicplayer/service/MusicService.java:38]: Useless import?

File: app/src/main/java/com/poupa/vinylmusicplayer/ui/activities/SettingsActivity.java
Comment (soncaokim) [inline app/src/main/java/com/poupa/vinylmusicplayer/ui/activities/SettingsActivity.java:29]: Useless import?

File: app/src/main/res/xml/pref_blacklist.xml
Comment (soncaokim) [reply app/src/main/res/xml/pref_blacklist.xml:12]: It would make more sense to me to do those changes.
It's really my personal preference though :-)

**Review by soncaokim — COMMENTED** [review]
> * (ie "only show content under the start folder: /storage/music/xxx")

Can you please add this, it would be helpful for the user

**Comment by soncaokim** [comment]
> SettingsActivity.java also implements OnSharedPreferenceChangeListener and seems a better place to intrigue the rescan.

I was thinking about a separate new class for this - LibraryFolderFilterBlahBlah (managing white and blacklisting).

It would require much more code to handle -> I'm okay with this shortcut, will get back to this if it still bother me.

**Comment by soncaokim** [comment]
> * I'd prefer merging the Setting's blacklist section (the existing blacklist and the new whitelist) into the Library section (on top).
>   These are options to control the content/presentation of the library.

Same here, either merge the different setting sections (they are for controlling the "library browsing" part of the app, they are not changing the behaviour of the "folder browsing" part of the app --> make it clear).

Or at least please change the "Black list" section title to something like "Library content filter"

File: app/src/main/java/com/poupa/vinylmusicplayer/service/MusicService.java
Comment (tramhao) [reply app/src/main/java/com/poupa/vinylmusicplayer/service/MusicService.java:38]: fixed.

File: app/src/main/java/com/poupa/vinylmusicplayer/ui/activities/SettingsActivity.java
Comment (tramhao) [reply app/src/main/java/com/poupa/vinylmusicplayer/ui/activities/SettingsActivity.java:29]: fixed.

**Comment by tramhao** [comment]
> > * (ie "only show content under the start folder: /storage/music/xxx")
> 
> Can you please add this, it would be helpful for the user

I'm checking on this and will update it later.

**Comment by tramhao** [comment]
Now because master upgraded, and my fork is not the same with master(I merged my other branck bluetooth-lyric), how can I resolve the simple confilct in this PR?  Thanks a lot! I've learned many things from actual coding.

**Comment by soncaokim** [comment]
> How can I resolve the simple confilct in this PR?

Not sure that I get your concern correctly. 

I'd start with the "Resolve conflicts" button below, and manually pick your change and inject them into the appropriate places in the new code structure.

**Comment by tramhao** [comment]
> > How can I resolve the simple confilct in this PR?
> 
> Not sure that I get your concern correctly.
> 
> I'd start with the "Resolve conflicts" button below, and manually pick your change and inject them into the appropriate places in the new code structure.

Thanks and that's exactly what I need. Before I have never used this function online as I thought I don't have write access to the master repo.

File: app/src/main/java/com/poupa/vinylmusicplayer/ui/activities/SettingsActivity.java
Comment (soncaokim) [inline app/src/main/java/com/poupa/vinylmusicplayer/ui/activities/SettingsActivity.java:152]: You can avoid copying the string literal here - instead use the `PreferenceUtil.WHITELIST_ENABLED` constant

Comment (tramhao) [reply app/src/main/java/com/poupa/vinylmusicplayer/ui/activities/SettingsActivity.java]: Thanks and I forgot this constant completely. Fixed.

**Review by soncaokim — COMMENTED** [review]
Overall this looks good to me.
If this is okay to everyone, let's do some real life testing before merge.

**Comment by tramhao** [comment]
> Overall this looks good to me.
> If this is okay to everyone, let's do some real life testing before merge.

Thanks for your patience:)

**Review by soncaokim — APPROVED** [review]

**Comment by soncaokim** [comment]
@AdrienPoupa  I'm merging this - supposing there is no debate on the new feature.
This helps me to avoid handling conflicts with other changes

**Pull request merged**
